[Distributed] Reimplement distributed call thunks completely in AST

ktoso opened the pull request:
Simplifies a ton of things but not complete yet.

ktoso committed:
[Distributed] dist actor always has default executor (currently)

ktoso committed:
[Distributed] extra test for missing makeEncoder

ktoso committed:
[DistributedDecl] Add DistributedActorSystem to known SDK types

ktoso committed:
[DistributedActor] ok progress on getting the system via witness

ktoso committed:
[Distributed] allow hop-to `let any: any X` where X is DistActor

xedin committed:
[Distributed] AST: Add an accessor to determine whether type is distributed actor

- Classes have specialized method on their declarations
- Archetypes and existentials check their conformances for
  presence of `DistributedActor` protocol.

xedin committed:
[Distributed] AST: Account for distributed members declared in class extensions

`getConcreteReplacementForProtocolActorSystemType` should use `getSelfClassDecl`
otherwise it wouldn't be able to find actor if the member is declared in an extension.

ktoso committed:
[Distributed] fix ad-hoc requirement checks for 'mutating'

[PreChecker] LookupDC might be null, so account for that

ktoso committed:
[Distributed] Completed AST synthesis for dist thunk

ktoso committed:
[Distributed][ASTDumper] print pretty distributed in right color in AST dumps

ktoso committed:
wip on making the local/remote calls

ktoso committed:
using the _local to mark the localCall as known local

ktoso committed:
[Distributed] fix passing Never when not throwing

ktoso committed:
fix lifetime of mangled string

ktoso committed:
[Distributed] Implement recordGenericSubstitution

ktoso committed:
[Distributed] Dont add .

ktoso committed:
[Distributed] dont emit thunk when func broken

ktoso committed:
[Distributed] fix tests; cleanups

ktoso committed:
[Distributed] cleanup, move is... funcs to DistributedDecl

ktoso committed:
[Distributed] Remove SILGen for distributed thunks, it is in Sema now!

ktoso commented:
@swift-ci please smoke test

ktoso committed:
[Distributed]  no need to check stored props in protocols

ktoso commented:
@swift-ci please smoke test

ktoso commented:
@swift-ci please build toolchain

ktoso committed:
remote not used flag

ktoso committed:
fix mangling test

ktoso commented:
@swift-ci please smoke test

ktoso commented:
@swift-ci please build toolchain

ktoso commented:
resolves rdar://84054772

ktoso committed:
cleanup

ktoso commented:
The failures are some linux only thing, we'll need to investigate...

xedin committed:
[Distributed] Synthesis: Don't re-use AST nodes for `decodeArgument` references

xedin committed:
[Distributed] Synthesis: Make sure that each thunk parameter has an internal name

xedin committed:
[Distributed/Synthesis] NFC: Add a comment regarding empty internal parameter names

xedin committed:
[Distributed] NFC: Adjust distributed thunk manglings in the accessor section test-cases

ktoso commented:
@swift-ci please smoke test and merge

ktoso commented:
@swift-ci please smoke test and merge

xedin commented on lib/Sema/CodeSynthesisDistributedActor.cpp:
```diff
@@ -81,9 +96,519 @@ static VarDecl *addImplicitDistributedActorIDProperty(
+/******************************************************************************/
+
+FuncDecl *GetDistributedThunkRequest::evaluate(
+    Evaluator &evaluator, AbstractFunctionDecl *afd) const {
```
Nit: A more descriptive name would be `thunk` instead of `afd`.

ktoso replied:
This is the "func" for which we make the "thunk" so I'll call it func I guess :)

xedin commented on lib/Sema/CodeSynthesisDistributedActor.cpp:
```diff
@@ -81,9 +96,519 @@ static VarDecl *addImplicitDistributedActorIDProperty(
+  if (!getConcreteReplacementForProtocolActorSystemType(afd)) {
+    // Don't synthesize thunks, unless there is a *concrete* ActorSystem.
+    // TODO(distributed): we should be able to lift this eventually,
+    // and allow resolving distributed actor protocols.
```
This has to emit an error message.

ktoso replied:
Hmm, we issue this request also for `distributed func` inside protocols and then we just return nothing I think.

Erroring here would cause us to reject

```
protocol NoSerializationRequirementYet: DistributedActor {
  // OK, no serialization requirement yet
  distributed func test() -> NotCodable
}
```

i think this protocol is fine 🤔 

We should reject such extension though I guess... That sounds right, I'll do that separately

xedin replied:
It sounds like the `if` here is actually an assertion and caller has to be detect and diagnose that something doesn't have a concrete actor system?

ktoso replied:
Yeah I think you're right, we should not call this in protocols I guess.

Will follow up with a patch for that. Then we can diagnose here and prevent bad extensions 👍

xedin replied:
Sounds good!

xedin commented on lib/Sema/TypeCheckConcurrency.cpp:
```diff
@@ -3050,10 +3110,10 @@ namespace {
-    /// a remote distributed actor in the given context.
-    bool isDistributedThunk(ConcreteDeclRef ref, Expr *context,
-                            bool isInAsyncLetInitializer);
+//    /// Determine whether the given reference is to a method on
```
Leftover comments?

xedin commented on lib/Sema/TypeCheckConcurrency.cpp:
```diff
@@ -4802,7 +4862,10 @@ bool swift::isPotentiallyIsolatedActor(
 
-  if (var->getName().str().equals("__secretlyKnownToBeLocal")) {
+  if (var->getName().str().equals("__secretlyKnownToBeLocal")
+      ) {
```
Leftover comments?

xedin commented on lib/SILGen/SILGenFunction.cpp:
```diff
@@ -522,6 +522,7 @@ void SILGenFunction::emitFunction(FuncDecl *fd) {
              fd->getResultInterfaceType(), fd->hasThrows(), fd->getThrowsLoc());
 
   if (fd->isDistributedActorFactory()) {
+    // TODO(distributed): generalize as "emit body synthesized in SIL" marker
```
I think this is outdated.

xedin commented on lib/Sema/TypeCheckConcurrency.cpp:
```diff
@@ -1903,6 +1912,45 @@ namespace {
+        if (auto conversion = dyn_cast<ImplicitConversionExpr>(expr))
+          expr = conversion->getSubExpr();
+
+//        // Map opaque values.
```
Leftover in-progress comment?

ktoso replied:
whoop yeah thank you

xedin commented on lib/Sema/TypeCheckConcurrency.cpp:
```diff
@@ -2809,6 +2867,7 @@ namespace {
 
       case ActorIsolationRestriction::CrossActorSelf:
       case ActorIsolationRestriction::ActorSelf:
+        valueRef.dump();
```
Leftover `dump()`

ktoso replied:
thx

xedin reviewed (approved):
Looks good to me! Majority of comments could be addressed separately only ones worth addressing now is missing mapTypeIntoContext and unnecessary copy of the thunk name.

ktoso commented on lib/Sema/CodeSynthesisDistributedActor.cpp:
```diff
@@ -30,6 +31,23 @@
+//  std::string localFuncNameString = "$dist_";
+  std::string localFuncNameString = "";
+  localFuncNameString.append(std::string(func->getBaseName().getIdentifier().str()));
+  auto thunkBaseName = DeclBaseName(C.getIdentifier(StringRef(localFuncNameString)));
```
right, thank you

xedin committed:
[Distributed] NFC: Adjust distributed thunk manglings in the accessor thunk test-cases

ktoso commented:
@swift-ci please smoke test

ktoso commented:
@swift-ci please smoke test linux

ktoso commented on lib/AST/DistributedDecl.cpp:
```diff
@@ -65,19 +65,50 @@ using namespace swift;
+    return signature->getConcreteType(ActorSystemAssocType);
+  }
+
+  llvm_unreachable("Unable to fetch ActorSystem type! Un");
```
thanks!

ktoso commented on lib/Sema/CodeSynthesisDistributedActor.cpp:
```diff
@@ -81,9 +96,519 @@ static VarDecl *addImplicitDistributedActorIDProperty(
+    forwardingParams.push_back(new (C) DeclRefExpr(
+        ConcreteDeclRef(param), DeclNameLoc(), /*implicit=*/true,
+        swift::AccessSemantics::Ordinary,
+        param->getInterfaceType()));
```
thanks!

ktoso commented on lib/Sema/CodeSynthesisDistributedActor.cpp:
```diff
@@ -81,9 +96,519 @@ static VarDecl *addImplicitDistributedActorIDProperty(
+  // --- invocationEncoder = system.makeInvocationEncoder()
+  auto *invocationVar =
+      new (C) VarDecl(/*isStatic=*/false, VarDecl::Introducer::Var, sloc,
+                      C.getIdentifier("invocation"), thunk);
```
we have them actually; will use 👍

ktoso commented on lib/Sema/CodeSynthesisDistributedActor.cpp:
```diff
@@ -81,9 +96,519 @@ static VarDecl *addImplicitDistributedActorIDProperty(
+    auto *makeInvocationArgs = ArgumentList::createImplicit(C, {});
+    auto makeInvocationCallExpr =
+        CallExpr::createImplicit(C, makeInvocationExpr, makeInvocationArgs);
+    makeInvocationCallExpr->setType(DTIE->getInterfaceType());
```
you're right, thanks!

ktoso committed:
review follow ups

ktoso committed:
xfail some linux tests for now so we can land the AST thunk

ktoso commented:
Addressed comments and XFAILed a few linux tests because I really want to land this PR, it's holding up a lot of the other ones.

Will fix the linux tests ASAP though, it is just mangling fixes i think.

ktoso commented:
@swift-ci please smoke test and merge

ktoso commented:
@swift-ci please smoke test and merge

ktoso commented:
@swift-ci please smoke test and merge

ktoso commented:
@swift-ci please smoke test

ktoso commented:
@swift-ci please smoke test

ktoso committed:
Update distributed_actor_remote_functions.swift

ktoso commented:
@swift-ci please smoke test
